File: Fundamentals_of_Software_Design/C_language/AVL_tree/avl_tree.c
#include <stdlib.h>
#include <stdio.h>

#include "avl_tree.h"

#define MAX(a, b) ((a) > (b) ? (a) : (b))

static BSTNode* create_node(Pointer data) {
    BSTNode* node = (BSTNode*)malloc(sizeof(BSTNode));
    if (!node) return NULL;
    node->data = data;
    node->left = NULL;
    node->right = NULL;
    node->height = 1;
    return node;
}

static int height(BSTNode* node) {
    return node ? node->height : 0;
}

static int balance_factor(BSTNode* node) {
    return node ? height(node->left) - height(node->right) : 0;
}

static void update_height(BSTNode* node) {
    if (node) {
        node->height = 1 + MAX(height(node->left), height(node->right));
    }
}

static BSTNode* rotate_right(BSTNode* y) {
    BSTNode* x = y->left;
    BSTNode* T2 = x->right;
    x->right = y;
    y->left = T2;
    update_height(y);
    update_height(x);
    return x;
}

static BSTNode* rotate_left(BSTNode* x) {
    BSTNode* y = x->right;
    BSTNode* T2 = y->left;
    y->left = x;
    x->right = T2;
    update_height(x);
    update_height(y);
    return y;
}

static BSTNode* balance_node(BSTNode* node) {
    if (!node) return node;
    update_height(node);
    int balance = balance_factor(node);
    if (balance > 1 && balance_factor(node->left) >= 0) {
        return rotate_right(node);
    }
    if (balance < -1 && balance_factor(node->right) <= 0) {
        return rotate_left(node);
    }
    if (balance > 1 && balance_factor(node->left) < 0) {
        node->left = rotate_left(node->left);
        return rotate_right(node);
    }
    if (balance < -1 && balance_factor(node->right) > 0) {
        node->right = rotate_right(node->right);
        return rotate_left(node);
    }
    return node;
}

static BSTNode* find_min(BSTNode* node) {
    while (node && node->left) {
        node = node->left;
    }
    return node;
}

static BSTNode* insert_node(BSTNode* node, Pointer data, CmpFunc cmp_func, Pointer* replaced_data) {
    if (!node) {
        return create_node(data);
    }
    int cmp_result = cmp_func(data, node->data);
    if (cmp_result < 0) {
        node->left = insert_node(node->left, data, cmp_func, replaced_data);
    }
    else if (cmp_result > 0) {
        node->right = insert_node(node->right, data, cmp_func, replaced_data);
    }
    else {
        *replaced_data = node->data;
        node->data = data;
        return node;
    }
    return balance_node(node);
}

static BSTNode* delete_node(BSTNode* node, Pointer data, CmpFunc cmp_func, Pointer* deleted_data) {
    if (!node) {
        return NULL;
    }
    int cmp_result = cmp_func(data, node->data);
    if (cmp_result < 0) {
        node->left = delete_node(node->left, data, cmp_func, deleted_data);
    }
    else if (cmp_result > 0) {
        node->right = delete_node(node->right, data, cmp_func, deleted_data);
    }
    else {
        *deleted_data = node->data;
        if (!node->left || !node->right) {
            BSTNode* temp = node->left ? node->left : node->right;
            if (!temp) {
                temp = node;
                node = NULL;
            }
            else {
                *node = *temp;
            }
            
            free(temp);
        }
        else {
            BSTNode* temp = find_min(node->right);
            node->data = temp->data;
            node->right = delete_node(node->right, temp->data, cmp_func, &temp->data);
        }
    }
    if (!node) {
        return node;
    }
    return balance_node(node);
}

static void clear_nodes(BSTNode* node) {
    if (!node) return;
    clear_nodes(node->left);
    clear_nodes(node->right);
    free(node);
}

static void foreach_inorder(BSTNode* node, void (*foreach_func)(Pointer data, Pointer extra_data), Pointer extra_data) {
    if (!node) return;
    foreach_inorder(node->left, foreach_func, extra_data);
    foreach_func(node->data, extra_data);
    foreach_inorder(node->right, foreach_func, extra_data);
}

static BSTNode* find_node(BSTNode* node, Pointer data, CmpFunc cmp_func) {
    if (!node) return NULL;
    
    int cmp_result = cmp_func(data, node->data);
    if (cmp_result < 0) {
        return find_node(node->left, data, cmp_func);
    }
    else if (cmp_result > 0) {
        return find_node(node->right, data, cmp_func);
    }
    else {
        return node;
    }
}

BST * bst_create(CmpFunc cmp_func) {
    BST* tree = (BST*)malloc(sizeof(BST));
    if (!tree) return NULL;
    tree->root = NULL;
    tree->cmp_func = cmp_func;
    tree->size = 0;
    return tree;
}

void bst_clear(BST *tree) {
    if (!tree) return;
    clear_nodes(tree->root);
    tree->root = NULL;
    tree->size = 0;
}

void bst_destroy(BST *tree) {
    if (!tree) return;
    bst_clear(tree);
    free(tree);
}

size_t bst_size(BST *tree) {
    return tree ? tree->size : 0;
}

Pointer bst_find(BST *tree, Pointer data) {
    if (!tree || !data) return NULL;
    BSTNode* node = find_node(tree->root, data, tree->cmp_func);
    return node ? node->data : NULL;
}

Pointer bst_insert(BST *tree, Pointer data) {
    if (!tree || !data) return NULL;
    Pointer replaced_data = NULL;
    tree->root = insert_node(tree->root, data, tree->cmp_func, &replaced_data);
    if (!replaced_data) {
        tree->size++;
    }
    return replaced_data;
}

Pointer bst_delete(BST *tree, Pointer data) {
    if (!tree || !data) return NULL;
    Pointer deleted_data = NULL;
    tree->root = delete_node(tree->root, data, tree->cmp_func, &deleted_data);
    if (deleted_data) {
        tree->size--;
    }
    return deleted_data;
}

void bst_foreach(BST *tree, void (*foreach_func)(Pointer data, Pointer extra_data), Pointer extra_data) {
    if (!tree || !foreach_func) return;
    foreach_inorder(tree->root, foreach_func, extra_data);
}

File: Fundamentals_of_Software_Design/Python_language/AVL_tree/avl_tree.py
class _Node:
    __slots__ = ('data', 'parent', 'left', 'right', 'height')
    def __init__(self, data):
        self.data = data
        self.parent = None
        self.left = None
        self.right = None
        self.height = 1

class Tree:
    __slots__ = ('root', 'cmp_func')
    def __init__(self, cmp_func):
        self.root = None
        self.cmp_func = cmp_func

def create(cmp_func):
    return Tree(cmp_func)

def clear(tree):
    tree.root = None

def size(tree):
    def _size(node):
        if node is None:
            return 0
        return 1 + _size(node.left) + _size(node.right)
    return _size(tree.root)

def _find_node(tree, data):
    node = tree.root
    while node is not None:
        cmp_result = tree.cmp_func(data, node.data)
        if cmp_result == 0:
            return node
        elif cmp_result < 0:
            node = node.left
        else:
            node = node.right
    return None

def find(tree, data):
    node = _find_node(tree, data)
    return node.data if node is not None else None

def _height(node):
    return node.height if node is not None else 0

def _update_height(node):
    if node is not None:
        node.height = 1 + max(_height(node.left), _height(node.right))

def _balance_factor(node):
    return _height(node.left) - _height(node.right) if node is not None else 0

def _rotate_left(tree, x):
    y = x.right
    x.right = y.left
    if y.left is not None:
        y.left.parent = x
    y.parent = x.parent
    if x.parent is None:
        tree.root = y
    elif x == x.parent.left:
        x.parent.left = y
    else:
        x.parent.right = y
    y.left = x
    x.parent = y
    _update_height(x)
    _update_height(y)
    return y

def _rotate_right(tree, y):
    x = y.left
    y.left = x.right
    if x.right is not None:
        x.right.parent = y
    x.parent = y.parent
    if y.parent is None:
        tree.root = x
    elif y == y.parent.right:
        y.parent.right = x
    else:
        y.parent.left = x
    x.right = y
    y.parent = x
    _update_height(y)
    _update_height(x)
    return x

def _rebalance(tree, node):
    while node is not None:
        _update_height(node)
        balance = _balance_factor(node)
        if balance > 1:
            if _balance_factor(node.left) < 0:
                node.left = _rotate_left(tree, node.left)
            node = _rotate_right(tree, node)
        elif balance < -1:
            if _balance_factor(node.right) > 0:
                node.right = _rotate_right(tree, node.right)
            node = _rotate_left(tree, node)
        node = node.parent

def insert(tree, data):
    new_node = _Node(data)
    if tree.root is None:
        tree.root = new_node
        return None
    
    current = tree.root
    parent = None
    while current is not None:
        parent = current
        cmp_result = tree.cmp_func(data, current.data)
        if cmp_result == 0:
            old_data = current.data
            current.data = data
            return old_data
        elif cmp_result < 0:
            current = current.left
        else:
            current = current.right
    
    cmp_result = tree.cmp_func(data, parent.data)
    if cmp_result < 0:
        parent.left = new_node
    else:
        parent.right = new_node
    new_node.parent = parent
    
    _rebalance(tree, parent)
    return None

def _min_node(node):
    current = node
    while current.left is not None:
        current = current.left
    return current

def delete(tree, data):
    node = _find_node(tree, data)
    if node is None:
        return None
    
    old_data = node.data
    
    if node.left is None and node.right is None:
        if node.parent is None:
            tree.root = None
        elif node == node.parent.left:
            node.parent.left = None
        else:
            node.parent.right = None
        _rebalance(tree, node.parent)
    
    elif node.left is None:
        if node.parent is None:
            tree.root = node.right
            node.right.parent = None
        elif node == node.parent.left:
            node.parent.left = node.right
            node.right.parent = node.parent
        else:
            node.parent.right = node.right
            node.right.parent = node.parent
        _rebalance(tree, node.parent)
    
    elif node.right is None:
        if node.parent is None:
            tree.root = node.left
            node.left.parent = None
        elif node == node.parent.left:
            node.parent.left = node.left
            node.left.parent = node.parent
        else:
            node.parent.right = node.left
            node.left.parent = node.parent
        _rebalance(tree, node.parent)
    
    else:
        successor = _min_node(node.right)
        node.data = successor.data
        if successor == successor.parent.left:
            successor.parent.left = successor.right
        else:
            successor.parent.right = successor.right
        if successor.right is not None:
            successor.right.parent = successor.parent
        _rebalance(tree, successor.parent)
    
    return old_data

def foreach(tree, func):
    def _inorder(node):
        if node is not None:
            _inorder(node.left)
            func(node.data)
            _inorder(node.right)
    _inorder(tree.root)
